Error when inline fallible non-terminal

I don't have the time to create a perfectly nice minimal example, sorry. But the failing code looks a bit like:

``` rust
ClassDecl: ast::TypeDef = {
    <NormalClassDecl> => ast::TypeDef::NormalClass(<>),
};

#[inline]
NormalClassDecl: ast::Class = {
    "something..." =>? { ... }
};
```

It results in 

```
error: mismatched types:
 expected `ast::item::Class`,
    found `core::result::Result<ast::item::Class, lalrpop_util::ParseError<...>>`
```

The relevant generated code is (with some cleanup):

``` rust
pub fn __action33<
>(
    __0: ast::Class,
    __lookbehind: &Option<BytePos>,
    __lookahead: &Option<(BytePos, Token, BytePos)>,
) -> ast::TypeDef
{
    ast::TypeDef::NormalClass(__0)
}

pub fn __action400<
>(
    __0: Token,
    __1: String,
    __2: Vec<ast::ClassMember>,
    __lookbehind: &Option<BytePos>,
    __lookahead: &Option<(BytePos, Token, BytePos)>,
) -> Result<ast::TypeDef,__ParseError<BytePos,Token,Report>>
{
    // this returns `Result<ast::Class,__ParseError<BytePos,Token,Report>>`
    let __temp0 = __action394(  
        __0,
        __1,
        __2,
        __lookbehind,
        __lookahead,
    );
    __action33(
        __temp0,     // <--- compiler error here
        __lookbehind,
        __lookahead,
    )
}


```

After brute force trying many different combinations of fallible/infallible branches, I have this minimal example which exposes two separate bugs:

### Minimal example

 ```rust
grammar;

pub Rule: i32 = {
    <i:Inlined> => i + 1337,
};

#[inline]
Inlined: i32 = "a" =>? Ok(420);
```

**Offending generated function**
```rust
fn __action3<   // The "inlining" action that composes the expressions `i + 1337` and `Ok(420)`
    'input,
>(
    input: &'input str,
    __0: (usize, &'input str, usize),    // <-- terminal "a"
) -> Result<i32,__lalrpop_util::ParseError<usize,Token<'input>,&'static str>>
{
    let __start0 = __0.0.clone();
    let __end0 = __0.2.clone();
    let __temp0 = __action2(   // <-- for the expr `Ok(420)`.  Returns Result<i32>
        input,
        __0,
    );
    let __temp0 = (__start0, __temp0, __end0);
    __action1(   // <-- for the expr `i + 1337`.  Returns i32
        input,
        __temp0,   // <-- expects (usize, i32, usize)
    )
}
```

**Errors**
```rust
    Checking lalrpop v0.19.4 (/mnt/f/asd/clone/lalrpop/lalrpop)
error[E0308]: mismatched types
   --> lalrpop/tests/min_0010.rs:372:9
    |
372 |         __temp0,
    |         ^^^^^^^ expected `i32`, found enum `std::result::Result`
    |
    = note: expected tuple `(_, i32, _)`
               found tuple `(_, std::result::Result<i32, __lalrpop_util::ParseError<usize, __lalrpop_util::lexer::Token<'_>, &str>>, _)`

error[E0308]: mismatched types
   --> lalrpop/tests/min_0010.rs:370:5
    |
370 | /     __action1(
371 | |         input,
372 | |         __temp0,
373 | |     )
    | |_____^ expected enum `std::result::Result`, found `i32`
    |
    = note: expected enum `std::result::Result<i32, __lalrpop_util::ParseError<usize, __lalrpop_util::lexer::Token<'input>, &'static str>>`
               found type `i32`
help: try using a variant of the expected enum
    |
370 |     Ok(__action1(
371 |         input,
372 |         __temp0,
373 |     ))
    |

error: aborting due to 2 previous errors
```

### Interpretation

There are two distinct errors here:

1.  The first error is because LALRPOP does not recognize that it needs to match on the result of a fallible action function from an inlined nonterminal.
2.  The second one happens because, even though LALRPOP correctly determines (most likely [due to this](https://github.com/lalrpop/lalrpop/blob/149e77b301828392b09da5c2ba634fd19ce3a3a7/lalrpop/src/normalize/inline/mod.rs#L101)) that `action3` should return a `Result`, it fails to recognize that it needs to `Ok`-wrap the infallible `__action1`.  (this error would not appear if `Rule` instead had a `=>?` pattern)

--- a/lalrpop-test/src/inline_fallible.lalrpop
+++ b/lalrpop-test/src/inline_fallible.lalrpop
@@ -1,0 +1,29 @@
+grammar;
+
+pub Inline: i32 =
+    <i:Inlined> => i + 400;
+
+pub MultipleInline: i32 =
+    <i:Inlined> <j:Inlined> => i + j;
+
+pub InlineIntoFallible: i32 =
+    <i:Inlined> =>? Ok(i + 400);
+
+pub ADifferentProductionIsFallible: i32 = {
+    <i:Inlined> => i + 400,
+    "b" =>? Ok(0),
+};
+
+pub RecursiveInline: i32 =
+    <i:RecursivelyInlined> => i + 400;
+
+#[inline]
+Inlined: i32 = {
+    "a1" =>? Ok(10),
+    "a2" => 20,
+};
+
+// to even further obscure the fallible-ness
+#[inline]
+RecursivelyInlined: i32 =
+    "c" <i:Inlined> "d" => 2 * i;

--- a/lalrpop-test/src/lib.rs
+++ b/lalrpop-test/src/lib.rs
@@ -110,6 +110,9 @@ lalrpop_mod!(
 /// test for inlining expansion issue #55
 lalrpop_mod!(issue_55);
 
+/// test for inlining of fallible NTs (issue #91)
+lalrpop_mod!(inline_fallible);
+
 /// test for unit action code
 lalrpop_mod!(unit);
 
@@ -839,6 +842,25 @@ fn issue_55_test1() {
     assert!(a.is_empty());
     assert_eq!(b, "Z");
     assert_eq!(c, vec!["X", "Y"]);
+}
+
+#[test]
+fn inline_fallible() {
+    assert!(inline_fallible::InlineParser::new()
+        .parse("a1")
+        .is_ok());
+    assert!(inline_fallible::MultipleInlineParser::new()
+        .parse("a2 a1")
+        .is_ok());
+    assert!(inline_fallible::InlineIntoFallibleParser::new()
+        .parse("a2")
+        .is_ok());
+    assert!(inline_fallible::ADifferentProductionIsFallibleParser::new()
+        .parse("a1")
+        .is_ok());
+    assert!(inline_fallible::RecursiveInlineParser::new()
+        .parse("c a2 d")
+        .is_ok());
 }
 
 #[test]

--- a/lalrpop/src/build/action.rs
+++ b/lalrpop/src/build/action.rs
@@ -382,7 +382,12 @@ fn emit_inline_action_code<W: Write>(
                     rust!(rust, "&{}start{},", grammar.prefix, temp_counter);
                     rust!(rust, "&{}end{},", grammar.prefix, temp_counter);
                 }
-                rust!(rust, ");");
+
+                if grammar.action_is_fallible(inlined_action) {
+                    rust!(rust, ")?;");
+                } else {
+                    rust!(rust, ");");
+                }
 
                 // wrap up the inlined value along with its span
                 rust!(
@@ -403,9 +408,19 @@ fn emit_inline_action_code<W: Write>(
             }
         }
     }
+
+    let final_action_fallible = grammar.action_is_fallible(data.action);
+    let (ok_begin, ok_end) = match (defn.fallible, final_action_fallible) {
+        (true, true) |
+        (false, false) => ("", ""),
+        (true, false) => ("Ok(", ")"),
+        (false, true) => unreachable!()
+    };
+
     rust!(
         rust,
-        "{}action{}{}",
+        "{}{}action{}{}",
+        ok_begin,
         grammar.prefix,
         data.action.index(),
         lparen
@@ -435,7 +450,7 @@ fn emit_inline_action_code<W: Write>(
         }
     }
     assert!(!data.symbols.is_empty());
-    rust!(rust, ")");
+    rust!(rust, "){}", ok_end);
 
     rust!(rust, "}}");
     Ok(())

